onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc

```diff
@@ -2006,10 +2020,31 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<Node*>& fuse
         }
       }
 
-      // Run TRT inference
-      if (!trt_context->enqueueV2(&buffers[0], stream, nullptr)) {
-        return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "TensorRT EP execution context enqueue failed.");
-      }
+      // Run TRT inference			  
+      if (trt_state->cuda_graph_enable)
+      {
+          if (*cuda_graph_ptr == nullptr) {
```

Comment:
Should we always capture the graph on the first Run() (when cuda_graph_ptr == nullptr) or should we allow the user to dictate that via a RunOption to state that they would like a particular Run to be captured ? If a graph had already been captured , we can reset it and capture a new graph at that point and this offers more flexibility to the user in terms of when they would like a new graph to be captured (useful for a dynamic shape model and the user would like to capture a new graph when they change the input shape). This is the idea that I had for the CUDA EP cuda graph support (https://github.com/microsoft/onnxruntime/pull/9978/) but it would be nice to land on a common design for all EPs...

Reply:
interesting idea. But would requiring users to set this on every Run() call be too much overhead and make it less user friendly?
I think the general use case is the user wants to enable graph capture wherever it's possible and launch graphs instead of individual kernels. 
maybe this requires more offline discussion.

Reply:
Well they can set a RunOption only when they would like a new graph to be captured ? During all other Runs, if a graph has been captured already, the graph is just replayed (the user doesn't need to pass in a flag here). This approach gives the user a choice to re-capture a graph for a new input within the same session. Otherwise, they would have to create a new session just to capture a new graph....

Reply:
ok i see. a way for the graph state to be reset manually by the user per Run().
is this only to handle case when the input shape changes?
i'm trying to think of a genuine production use case. graph capture introduces a lot of overhead. 
would the user be resetting the graph frequently in a normal production workflow? i would think they would only ever want to do it once for the session. maybe for an interactive/debugging type session?

Reply:
We should avoid any frequent graph capturing (due to input shape changes) in the inference. If users know the input shape will never change in a inference run but want to change input shape for different run, putting cuda_graph in RunOptions will save the users session creation time. But it seems more likely for a debugging session. In production, the input shape are either static or dynamic for all inference runs without user's interaction.

Reply:
Ok, seems fair to wait until there is a such an ask before adding the capability.